rfc/drafts/indexer-ideas-v1.md

```diff
@@ -0,0 +1,303 @@
+# Idea Sketches:
+
+Rough sketches of some high-level ideas. Some sections to be re-drafted into RFCs. Contents:
+
+- General questions & ideas.
+- Various term definitions.
+- WIP outline for the V1 of search & dedupe architecture(s).
+- Beginnings of iteration roadmap toward distributed search & dedupe.
+
+
+## GENERAL MC QUESTIONS / IDEAS:
+
++ In addition to edges that support a relationship, also support edges that deny a relationship?
+  - E.g. `NotArtefactDerivationCell`, `NotArtefactCreationCell`, etc.
+  - Or, adding an "invert_relationship" field for the EntityLinkCell. Not to be confused with
+    inverting the direction of the relationship.
+  
++ Support a denser format for specifying edges, other than just one pair at a time?
+  - E.g. `EntityLinkCell = {"group_entity_link":[(ref1,ref2,ref3),(ref4,ref5),(ref6)]}`
+  - Meaning that all grouped entities in that list should have the indicated relationship, and all separated
+    entities should have the inverse relationship. I.e. is_derivation vs is_not_derivation.
+  - Could be easily converted to all-pairs of links internally.
+
++ Object-localization for video / audio / images?
+  - System identifies bounding-boxes for multiple objects in an image / time ranges for multiple known audio clips
+    in an audio broadcast / time ranges for known video clips.
+  - System allows user to focus on just one of those detected objects in their search / dedupe query.
+  - Relevant to the search / dedupe architecture, because this likely would require for those multiple objects
+    to be indexed separately, along with their location in the media.
+
++ Non-monetary incentives for resource providers? I.e. User donates resources because he likes particular content.
+  - Force resource providers to support all content equally (Freenet).
+  - Allow full control over the content you're supporting (Bittorrent).
+  - Something in the middle (Tor exit nodes)?
+
++ Resource usage considerations?
```

Comment:
This is an important but out of scope discussion: let's assume fully cooperating peers where we have choice of shard allocations for now. Broadly speaking, what we want is a pattern of contributions that mirrors real-world relationships, e.g. indexers of art from public museums might cooperate etc.